### PR TITLE
perf(speculative): cache the EAGLE-3 draft-vocab token map across runs

### PR DESCRIPTION
**Pull request by khazic**
## What

EAGLE-3 builds its draft vocabulary by scanning the **entire training set** once
per setup — `build_eagle3_token_mapping` iterates the whole dataloader to rank
token frequencies and pick the top-`draft_vocab_size` ids. On large datasets that
scan dominates startup and is repeated identically on every run.

This PR adds an optional `recipe_args.selected_token_ids_path`. When set, the
first run writes the selected token ids there and later runs load them and skip
the scan; when unset, behavior is unchanged. The mapping is resolved through a
new `load_or_build_eagle3_token_mapping` orchestrator in
`components/datasets/llm/eagle3.py`:

- **save** is atomic (`.tmp` + `os.replace`) and stores only the ids plus
  `target_vocab_size` — the boolean mask is derived on load.
- **load** validates `target_vocab_size` and the resulting draft-vocab size,
  returning `None` (→ rebuild) on any mismatch or unreadable file, with a logged
  reason.
- **under DDP**, only rank 0 reads the file and broadcasts the ids to all ranks,
  so every rank takes the same load-vs-build branch. This is required because
  `build_eagle3_token_mapping` issues a collective `all_reduce` that would
  deadlock if ranks diverged, and it stays correct even on node-local
  (non-shared) filesystems.

## Changelog

- Add optional `recipe_args.selected_token_ids_path` to the EAGLE-3 recipe: cache
  the draft-vocab token map and skip the full-dataset frequency scan on reruns.
- Add `save_eagle3_token_mapping` / `load_eagle3_token_mapping` /
  `load_or_build_eagle3_token_mapping` to `components/datasets/llm/eagle3.py`.
- Document the knob in `examples/speculative/eagle3/llama_eagle3_mvp.yaml`.

## Design notes

- The cache is keyed only on the vocab sizes, **not** a dataset fingerprint: it is
  an explicit, user-controlled path, documented to be repointed (or deleted) when
  the training data changes. This keeps the perf feature from re-hashing the
  dataset (which would defeat the point of skipping the scan).
- Complementary to the existing `eagle_meta.pt` mechanism: on **resume** the
  checkpoint's saved mapping still overrides this cache, so `selected_token_ids_path`
  only affects **cold starts**.

## Pre-checks

- [x] `ruff format` / `ruff check` clean on changed files.
- [x] Unit tests (`tests/unit_tests/speculative/test_eagle3_token_cache.py`):
  save/load round-trip, incompatible/corrupt-cache rebuild paths, orchestrator
  scan-skip on a cache hit, and a single-process gloo run exercising the rank-0
  load + broadcast path. EAGLE suites: 85 passed, 2 skipped (GPU-gated).
- [x] DCO sign-off present.

### CODE REVIEW

File: nemo_automodel/components/datasets/llm/eagle3.py [hunk 1]
```diff
@@ -162,3 +166,158 @@ def build_eagle3_token_mapping(
...
+    vocab size -- it does NOT fingerprint the dataset or tokenizer. A cache built
+    from a different dataset still loads cleanly, so a caller that changes the
+    training data must point ``selected_token_ids_path`` at a fresh location (or
+    delete the file). Returns ``None`` -- so the caller rebuilds -- when the file
+    is missing, unreadable, or its stored vocab sizes do not match the config.
+    """
+    if not os.path.exists(path):
+        return None
+    try:
+        data = torch.load(path, map_location="cpu", weights_only=False)
```
[the remark this answers is not in the document]
Comment (khazic) [reply nemo_automodel/components/datasets/llm/eagle3.py]: Adopted in e1e40291 — switched to `weights_only=True`. The payload is only `{selected_token_ids: Tensor, target_vocab_size: int}`, both supported by the safe loader on PyTorch 2.6+ (project minimum), and the existing try/except still falls through to a rebuild on any load failure, so it's a no-op for valid caches and hardens the user-controlled path. Round-trip unit tests pass.

One correction for the record: the ruff S301/S403 rationale doesn't actually apply here — S301 (`suspicious-pickle-usage`) only matches the `pickle` module, not `torch.load`, and S403 is selected but inert (`has no effect because preview is not enabled`), so `ruff check --select S301,S403` on this file passes. The change is still worth making on its own merits (defense-in-depth on a load of a trivially-safe payload), just not because ruff flags it.

**Review by HuiyingLi — APPROVED** [review]
